src/Microsoft.AspNetCore.Blazor.Server/Circuits/BlazorHub.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Blazor.Services;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Blazor.Server.Circuits
+{
+    internal class BlazorHub : Hub
+    {
+        private static readonly object CircuitKey = "circuit";
+
+        private readonly CircuitFactory _circuitFactory;
+
+        public BlazorHub(CircuitFactory circuitFactory)
+        {
+            _circuitFactory = circuitFactory ?? throw new ArgumentNullException(nameof(circuitFactory));
+        }
+
+        public CircuitHost CircuitHost
+        {
+            get => (CircuitHost)Context.Items[CircuitKey];
+            set => Context.Items[CircuitKey] = value;
+        }
+
+        public override Task OnDisconnectedAsync(Exception exception)
+        {
+            CircuitHost.Dispose();
+            return base.OnDisconnectedAsync(exception);
+        }
+
+        public async Task StartCircuit(string uriAbsolute, string baseUriAbsolute)
+        {
+            var circuitHost = _circuitFactory.CreateCircuit(Context.GetHttpContext(), Clients.Caller);
```

Comment:
Would `CreateCircuitHost` be a clearer name?

Reply:
sure